src/xmppjs/HistoryManager.ts

```diff
@@ -0,0 +1,102 @@
+import { Element } from "@xmpp/xml";
+import { JID } from "@xmpp/jid";
+
+export interface IHistoryLimits {
+    maxChars?: number | undefined,
+    maxStanzas?: number | undefined,
+    seconds?: number | undefined,
+    since?: Date | undefined,
+}
+
+/**
+ * Abstraction of a history storage backend.
+ */
+interface IHistoryStorage {
+    // add a message to the history of a given room
+    addMessage: (chatName: string, message: Element, jid: JID) => Promise<void>;
+    // get the history of a room.  The storage should apply the limits that it
+    // wishes too, and remove the limits that it applied from the `limits`
+    // parameter.  The returned list of Elements must include the Delayed
+    // Delivery information.
+    getHistory: (chatName: string, limits: IHistoryLimits) => Promise<Element[]>;
+}
+
+// pad a number to be two digits long
+function pad2(n: number) {
+    return n < 10 ? "0" + n.toString() : n.toString();
+}
+
+/**
+ * Store room history in memory.
+ */
+export class MemoryStorage implements IHistoryStorage {
+    private history: Map<string, Element[]>;
+    constructor(public maxHistory: number) {
+        this.history = new Map();
+    }
+
+    async addMessage(chatName: string, message: Element, jid: JID): Promise<void> {
```

Comment:
This function does not need to be async.
You may choose to make it async if planned features likely require it to become async.